### PR TITLE
Migration process started

### PR DESCRIPTION
**Pull request by balage1551**
Hi Stefan,

I've just started to analyze the two versions. I havn't made any changes, but created a document (migration checklist) where I'll list the tasks and decision points about the merge of the two versions.

This pull request is only for creating a place to discuss the issues and make the decisions.

I'll extend the MigrationChecklist as I dig my way through the code. 
Now there are three issues we have to make conclusions, then I can move on. It is in the file, but also I copy it below for easier reference:

Balage

### Task 1: Capacity to SizeDimension

**<u>Nature:</u>** 

`Capacity` became deprecated but kept, it is fully compatible with the `SizeDimension` class, which is a subclass of `Capacity`.

**<u>Decision:</u>** 

Make `Capacity` *deprecated*. Keep both classes.

**<u>State:</u>**

Ready.

**<u>Migration task:</u>**

None required. Optionally, but strongly recommended to replace all `Capacity` references by `SizeDimension`. (Only a search and replace.)

### Task 2: The Job interface

**<u>Nature:</u>** 

The `Job` interface was extended. This would break any third party implementation of this interface (or the extension of the `AbstractJob` abstract class).

**<u>Decision:</u>** 

There are several ways to overcome on this compatibility conflict:

1. Keep it as it is and force the implementators to extend its implementation or to migrate to `CustomJob`.
2. Restore the original interface and create an extended one as a descendant of this. The `AbstractJob` is the only one implementing this interface, so it can simply implement the extended one.

Option 1 is a compatibility breaker, but doesn't pollute the code unnecessary and urges the users to switch to new job structure. With Option 2, we are keeping compatibility, but introducing an unnecessary level in the class hierarchy and finding a good name for the extended interface would be hard, because this is only created for the sake of compatibility.

**<u>State:</u>**

(!) Waiting for decision.

**<u>Migration task:</u>**

*Write down detailed description of all mandatory and optional migration tasks.*

### Task 3: The old Job types

**<u>Nature:</u>** 

There are old job types (`Service`, `Pickup`, `Delivery` and `Shipment`) with deprecated methods. With the new `CustomJob` and the special functions (`addService`, `addPickup`, etc.) of its builder, these old job types are redundant and they are only special cases for the more general `CustomJob`. However, they are the central building classes for Jsprit, so keeping and supporting them in some way is vital for backward compatibility.

**<u>Decision:</u>** 

There are several ways to overcome on this compatibility conflict:

1. Remove these classes and guide the users to migrate to the methods in `CustomJob.Builder`.
2. There are already implementations for these classes in new version, but they are not compatible with the old once. Help the users to migrate to these classes.
3. Create façade classes. These classes offer the same interface as the old ones, but extends `CustomJob` and delegates the calls to the new structure.

The first lefts the least of the garbage in the code base, but seems to be too drastic and puts too much migration work on the old users. The second is the way how now the new version works, so here mostly documentation tasks are necessary. The third one would provide an *almost* 100% compatibility, but for a cost of one more redirection, which may have (although possibly minor) performance impact on Jsprit. The *almost* is there, because there may be a few method name collisions where both `CustomJob` and the old classes have the same method, but with different behavior.

My proposal is the 3rd one.

**<u>State:</u>**

(!) Waiting for decision.

**<u>Migration task:</u>**

*Write down detailed description of all mandatory and optional migration tasks.*

### CODE REVIEW

**Comment by balage1551** [comment]
The other document in the commit is only the first piece of the documentation for the new version. This was also the file I've got myself used to the markup syntax and to test several editors (Typora is the clear winner: easy, WYSIWYG, free).

**Comment by oblonski** [comment]
@balage1551 Thanks a lot.

Task 2: I prefer option 1 also. If we describe how to migrate it should not be a big problem.
Task 3: I aggree with your proposal, 3rd option.

File: docs/2.0/IntroductionToCustomJobs.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+# Introduction to Custom Jobs
+
```
Comment (oblonski) [inline docs/2.0/IntroductionToCustomJobs.md]: Perfect. Thanks.

**Comment by oblonski** [comment]
@balage1551 Should I merge the latest master into graphhopper:job-activity-refactoring?

**Comment by balage1551** [comment]
I don't think it is a good idea before we do all the migration preparations. Now there are several uncompatibilities I hope to be able to remove or facade. But you know better what was changed in the last half year and how much impact do they have on the job/ability structure.
Let's first alter the job-activity-refactoring, then do the merge.

**Comment by balage1551** [comment]
Today I did some huge steps on the migration. (See in https://github.com/balage1551/jsprit ?)

(1) I merged the legacy and new Job interfaces and AbstractJob classes. Happily, there were no such conflicts I couldn't resolve. So now the facade of the legacy implementation implements the new one while still compatible with the old API. This is a great achivement.

(2) After a lot of thinking, I made the decision about the old and new special job types. First I thought to completely remove the new Service, Pickup, Delivery and Shipment classes, because now they can be easily emulated by the CustomJob.Builder. However, it would need to rewrite almost all the tests and some of the main code to use the CustomJob. On the other hand, I couldn't keep the classes as they are, because for backward compatibility, the legacy facades should be placed into the same directory. The solution is: I named the new job from Service to ServiceJob and now the legacy one could be Service. 

(3) Created a documentation for migration and started to add all todos as they turn up. (https://github.com/balage1551/jsprit/blob/master/docs/2.0/MigrationGuide.md ?)

Now the first step is taken and looks right at the first time.
Let's discuss it, and if you like it, I'll make the same for the other three job classes. Now it would be an easy task.

**Comment by balage1551** [comment]
Now we reached a milestone.

The three original job types have got facaces and simultanously matches the API of the legacy and the new interfaces. The makes the old implementations almost compatible with the new ones. 

(Interestingly, three real problem tests fails with different cost values, but they failed even before the merge.)

Also I made (started is a better expression) the documentation for both guiding the migration and uderstanding the new concepts.

I think, at last, the time has come when your merge may be taken place. I am looking forward to it. Let me know if I may be in any help!

What would be the next step? Until I could give more help, I'll start to review, update, and correct the java doc of the new version. _It's a dirty job, but somebody's gotta do it._ ;-)

**Comment by oblonski** [comment]
> I think, at last, the time has come when your merge may be taken place. I am looking forward to it. Let me know if I may be in any help!

@balage1551 First of all, thanks a lot for your effort. Looks good at the first sight. Does this (aboove) mean that I can merge the latest master to this branch?

**Comment by oblonski** [comment]
> What would be the next step?

- migrate all our projects to jsprit v2 (with your migration guide)
- identify and solve our migration problems

**Comment by balage1551** [comment]
@oblonski "Does this (aboove) mean that I can merge the latest master to this branch?" I think, yes. The meantime, I complete the documentation and help when any migration issue turns up. 

If you run into any issue, which is not documented, let me know and I upgrade the migration guide.

**Comment by balage1551** [comment]
While I've documented the Jobs and activities I've found some small issues I could fix, but would like to hear your opinion.

**1. The factory and copy methods.** 
In the original code there are several classes with newInstance methods. They do nothing else, but calls the constructors with the same parameters. I don't see the purpose of it. Neither of the reason to create these static methods stands: we don't make any preparation before instantiation, there is no parameter collision in the overloaded constructors, nor there is a case when the concrete implemetation class is or will have to be hidden in the future. 
I didn't add these methods to the new classes, but it is there in some of the old once. For consistency, _I propose to remove these functions (a) from everywhere_ (this may give one more backward incompatibility, but mostly in internal and deprecated classes and is easy to migrate), _(b) or from internal classes_: the ones most probably the users rarely or never instantiated.

**2. single time window**
There are some occurancies where the code still calls the getSingleTimeWindow() function. It is for backward compatibility, but fragile and most probably indicates code where multiple time window support is not introduced. The only exception is the Break, where there is always a single time window. 
_I propose to get rid of the this function._

**3. isShipment function**
There is an isShipment function which has not much meaning now, so we _should eliminate it_. However, it is in some code I don't really understand yet, so I have to be careful and there may be case where I would have to ask your help.

**Comment by balage1551** [comment]
@oblonski As I try to eliminate singleTimeWindow, and isShipment I ran into several problems and questions:
1. With multiple time windows available, what should be set as intital _Theoritical Earliest_ and _Latest Start time_? The first time window? (If so, I would remove references to single time window and would move this code to the TimeWindows, so it could serve this request.)
2. Am I right, that vehicleRoute is a class only internally (by the algorithm) created? If I'm right, it can be safely convert to the new version without any backward compatibility overhead.
3. The JobActivityFactory (and its implementations) is now seem deprecated and needless (now the Jobs are responsible to create their activities) and may (have to) be removed. I'll do it.
4. We should eliminate any dependancy to shipments. It is now not the sole multi-activity job, so where it appears either the dependancy should be extended to any multi-activity jobs or should be eliminated entirely. I'll collect where it is used and you may tell me what to do.

**Comment by oblonski** [comment]
@balage1551 I a have a few unit tests that failed. Do you have any idea of the reasons?

**Comment by oblonski** [comment]
It seems to be related to the modulare obj function. If I replace it with the current one, it works.

**Comment by balage1551** [comment]
I'll compare the two versions today to find out where I made the mistake.

**Comment by oblonski** [comment]
👍

**Comment by oblonski** [comment]
@balage1551 Found this in AbstractJob

 ```
/**
     * Sets the index of the job within the problem.
     * <p>
     * <b>This method isn't part of the public API and should not be called!</b>
     * </p>
     *
     * @param index
     *            The index.
     */
    public void impl_setIndex(int index) {
        this.index = index;
    }
```

If this is not part of the public API (which is important and correct) why did you declare it public?

**Comment by balage1551** [comment]
@oblonski "If this is not part of the public API (which is important and correct) why did you declare it public?"
This is a workaround solution for the weakness of the Java visibility mechanism. In C++ there is a "friend" relationship, where you can declare which classes can interact with your protected API. In Java, you have only public, protected or package visibility. If you have a method which have to be reached from another (non-descendant) class of an other package, you can't do it. In these cases, this is the pattern most of the APIs use: declare it public (because you must), but mark it something not meant to be called by users. Using the "impl_" prefix is a widely used way to mark it. (I met with this and learnt from several complex frameworks, like JavaFX, Android ADK, etc.) This methods are not part of the public API, therefore you make no commitment of its stability, existence or behaviour.

In this case, we would like to keep the jobs quasi-imutable, but the index can't be set in construction time. Therefore we have to let it set later and it can be done only by publishing this method.

**Comment by oblonski** [comment]
Even I have some headache leaving setIndex public to the user (since this is strictly internal stuff) ...

**Comment by balage1551** [comment]
I don't know. 
It isn't seems right to reorganize code (separate AbstractJob from the job package) in an unnatural way just to hide a method.
As far as design concepts go, the pritiest solution would be to create a wrapper class containing an AbstractJob and an index and wrapping the jobs. Then delegating all methods of the AbstractJob. This would keep the Jobs (as they are seen from outside) imutable and without the need of a public setIndex. It isn't hard but has its own cost. The question is: does it worth?

**Comment by balage1551** [comment]
Well, I've found a very [nice trick](https://stackoverflow.com/questions/182278/is-there-a-way-to-simulate-the-c-friend-concept-in-java) to simulate "friend" class relationship.
It's simple, elegant and we don't have to force unnatural code organizing.

**Comment by oblonski** [comment]
> 1. The factory and copy methods.
> In the original code there are several classes with newInstance methods. They do nothing else, but calls the constructors with the same parameters. I don't see the purpose of it.

If the constuctor is public, there is indeed no purpose. Therefore, we can remove the factory method. Originally, I just followed the advice of Joshua Bloch who described some advantage of static factory methods over using constructors: http://www.informit.com/articles/article.aspx?p=1216151.

**Comment by oblonski** [comment]
> 2. single time window
> There are some occurancies where the code still calls the getSingleTimeWindow() function. It is for backward compatibility, but fragile and most probably indicates code where multiple time window support is not introduced. The only exception is the Break, where there is always a single time window.
> I propose to get rid of the this function.

Ok. I aggree.

**Comment by oblonski** [comment]
> 3. isShipment function
> There is an isShipment function which has not much meaning now, so we should eliminate it. However, it is in some code I don't really understand yet, so I have to be careful and there may be case where I would have to ask your help.

I aggree too. We should find out where we and why we use it and find another solution. If we want to the user to be able to specify arbitrary jobs, we should find alternatives to relying on hard-coded isShipments check and generalize.

**Comment by balage1551** [comment]
> If the constuctor is public, there is indeed no purpose. Therefore, we can remove the factory method. Originally, I just followed the advice of Joshua Bloch who described some advantage of static factory methods over using constructors: http://www.informit.com/articles/article.aspx?p=1216151.

I know Joshua Bloch's book well. ;-) But here neither of the advantages stand: we don't use it to give name, we don't want to use cached objects and we don't want to create subtypes of objects. On the other hand, its disadvantages stand. So I propose to make it uniform and drop the getInstances (except in the legacy code).

1. OK. I'll do it.
2. Ok. I'll do it too.
3. Ok. I'll do where I can. :-)

**Comment by oblonski** [comment]
Thanks, @balage1551 

> 3. Ok. I'll do where I can. :-)

Let me do this :).

**Comment by balage1551** [comment]
> Let me do this :).

OK. This is related with the 4th point of my second list of comments.

> We should eliminate any dependancy to shipments. It is now not the sole multi-activity job, so where it appears either the dependancy should be extended to any multi-activity jobs or should be eliminated entirely. I'll collect where it is used and you may tell me what to do.

So I leave it for you, too.

**Comment by oblonski** [comment]
Perfect. Thanks.

**Pull request merged**